crates/client/rpc/src/versions/v0_8_0/methods/ws/lib.rs

```diff
@@ -0,0 +1,398 @@
+                    return WsResult::Err(StarknetWsApiError::internal(Cow::from(format!(
+                        "Failed to retrieve block info for block {block_n}",
+                    ))));
+                };
+
+                if block_n < block_info.header.block_number.saturating_sub(BLOCK_PAST_LIMIT) {
+                    return WsResult::Err(StarknetWsApiError::TooManyBlocksBack);
+                }
+
+                block_n
+            }
+            starknet_core::types::BlockId::Hash(block_hash) => {
+                let Ok(Some(block_info)) = self.backend.get_block_info_from_block_latest() else {
+                    return WsResult::Err(StarknetWsApiError::internal(Cow::from(format!(
+                        "Failed to retrieve block info at hash {block_hash:#x}",
+                    ))));
+                };
+
+                let Ok(Some(block_n)) = self.backend.block_hash_to_block_n(&block_hash) else {
+                    return WsResult::Err(StarknetWsApiError::internal(Cow::from(format!(
+                        "Failed to retrieve block info at hash {block_hash:#x}"
+                    ))));
+                };
+
+                if block_n < block_info.header.block_number.saturating_sub(BLOCK_PAST_LIMIT) {
+                    return WsResult::Err(StarknetWsApiError::TooManyBlocksBack);
+                }
+
+                block_n
+            }
+            starknet_core::types::BlockId::Tag(starknet_core::types::BlockTag::Latest) => {
+                let Ok(Some(block_n)) = self.backend.get_latest_block_n() else {
+                    return WsResult::Err(StarknetWsApiError::internal(Cow::from(
+                        "Failed to retrieve block info for latest block",
+                    )));
+                };
+
+                block_n
+            }
+            starknet_core::types::BlockId::Tag(starknet_core::types::BlockTag::Pending) => {
+                return WsResult::Err(StarknetWsApiError::internal(Cow::from(
+                    "`starknet_subscribeNewHeads` does not support pending blocks",
+                )))
+            }
+        };
+
+        let mut rx = self.backend.subscribe_block_info();
+        for n in block_n.. {
+            if sink.is_closed() {
+                return WsResult::Ok;
+            }
+
+            let block_info = match self.backend.get_block_info_from_block_n(n) {
+                Ok(Some(block_info)) => block_info,
+                Ok(None) => break,
+                Err(e) => {
+                    return WsResult::Err(StarknetWsApiError::internal(Cow::from(format!(
+                        "Failed to retrieve block info for block {n}: {e}"
+                    ))))
+                }
+            };
+
+            let res = send_block_header(&sink, block_info, block_n).await;
+            if matches!(res, WsResult::Err(_)) {
+                return res;
+            }
+
+            block_n += 1;
```

Comment:
comment why this addition cannot overflow even though it's a user controlled input?

Reply:
No this is a very good point, this would actually be possible with the way I've currently set up the function.